src/Datadog.Trace/Containers/ContainerInfo.cs

```diff
@@ -0,0 +1,118 @@
+        /// When reading from file system, we only read one line at a time until we find a match,
+        /// which is usually the first line.
+        /// </remarks>
+        /// <param name="contents">The contents of a cgroup file.</param>
+        /// <returns>The container id if a match is found; otherwise, <c>null</c>.</returns>
+        public static string ParseCgroupText(string contents)
+        {
+            if (contents == null)
+            {
+                return null;
+            }
+
+            IEnumerable<string> lines = SplitLines(contents);
+            return ParseCgroupLines(lines);
+        }
+
+        /// <summary>
+        /// Used by <see cref="ParseCgroupText"/> to split the entire file contents
+        /// into individual lines to mock <see cref="File.ReadLines(string)"/>.
+        /// </summary>
+        /// <param name="contents">The multi-line string.</param>
+        /// <returns>An enumerable that returns each line from <paramref name="contents"/> when iterated.</returns>
+        public static IEnumerable<string> SplitLines(string contents)
+        {
+            if (contents == null)
+            {
+                yield break;
+            }
+
+            using (var reader = new StringReader(contents))
+            {
+                while (true)
+                {
+                    string line = reader.ReadLine();
+
+                    if (line == null)
+                    {
+                        yield break;
+                    }
+
+                    yield return line;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Uses regular expression to try to extract a container id from the specified string.
+        /// </summary>
+        /// <param name="lines">Lines of text from a cgroup file.</param>
+        /// <returns>The container id if found; otherwise, <c>null</c>.</returns>
+        public static string ParseCgroupLines(IEnumerable<string> lines)
+        {
+            return lines.Select(ParseCgroupLine)
+                        .FirstOrDefault(id => !string.IsNullOrWhiteSpace(id));
+        }
+
+        /// <summary>
+        /// Uses regular expression to try to extract a container id from the specified string.
+        /// </summary>
+        /// <param name="line">A single line from a cgroup file.</param>
+        /// <returns>The container id if found; otherwise, <c>null</c>.</returns>
+        public static string ParseCgroupLine(string line)
+        {
+            var lineMatch = Regex.Match(line, ContainerIdRegex);
+
+            return lineMatch.Success
+                       ? lineMatch.Groups[1].Value
+                       : null;
+        }
+
+        private static string GetContainerIdInternal()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && File.Exists(ControlGroupsFilePath))
+            {
+                IEnumerable<string> lines = File.ReadLines(ControlGroupsFilePath);
```

Comment:
Since `File.ReadLines` can throw exceptions, we should try/catch here. Related to Colin's comment, do we have a standard across tracers to indicate failure to get the container name?

Reply:
I added exception handling and logging in d78a4cc1f7a92714e4c4aabef45c5dd7a086c4f5.